### ISSUE TITLE
Train the real and fake data separately or simultaneously？

### ISSUE THREAD

**Issue by LynnHo** [AlexiaJM/Deep-learning-with-cats#5]
Thanks for the good code!!!

Can anybody explain the difference between training  the real and fake date separately and training them simultaneously. The code  backwards the real first then the fake.

**Comment by AlexiaJM** [comment]
What happens is we initialize the gradients as 0 with "D.zero_grad()", then we add up the gradients of the true examples when using backward(), then we add up the gradients of the fake examples when using backward() and finally we use the optimizer to update the parameters in the direction of the gradients. So backward() doesn't do the learning, it just adds the gradients to the variables (since variables store both data (.data) and gradients (.grad)).

See autograd documentation: http://pytorch.org/docs/master/autograd.html?highlight=backward#torch.autograd.backward
